src/main/java/edu/kit/kastel/state/ActiveAssessment.java

```diff
@@ -0,0 +1,132 @@
+import com.intellij.openapi.editor.markup.HighlighterLayer;
+import com.intellij.openapi.editor.markup.HighlighterTargetArea;
+import com.intellij.openapi.editor.markup.RangeHighlighter;
+import com.intellij.openapi.editor.markup.TextAttributes;
+import com.intellij.ui.JBColor;
+import edu.kit.kastel.extensions.settings.ArtemisSettingsState;
+import edu.kit.kastel.sdq.artemis4j.ArtemisClientException;
+import edu.kit.kastel.sdq.artemis4j.grading.Annotation;
+import edu.kit.kastel.sdq.artemis4j.grading.Assessment;
+import edu.kit.kastel.sdq.artemis4j.grading.ClonedProgrammingSubmission;
+import edu.kit.kastel.sdq.artemis4j.grading.penalty.GradingConfig;
+import edu.kit.kastel.sdq.artemis4j.grading.penalty.MistakeType;
+import edu.kit.kastel.utils.ArtemisUtils;
+import edu.kit.kastel.utils.CodeSelection;
+import edu.kit.kastel.utils.EditorUtil;
+import edu.kit.kastel.wrappers.AnnotationWithTextSelection;
+
+public class ActiveAssessment {
+    private final List<Consumer<List<AnnotationWithTextSelection>>> annotationsUpdatedListener = new ArrayList<>();
+
+    private final List<AnnotationWithTextSelection> annotations;
+    private final Assessment assessment;
+    private final ClonedProgrammingSubmission clonedSubmission;
+
+    public ActiveAssessment(Assessment assessment, ClonedProgrammingSubmission clonedSubmission)
+            throws IOException, ArtemisClientException {
+        this.assessment = assessment;
+        this.clonedSubmission = clonedSubmission;
+        this.annotations = new ArrayList<>(assessment.getAnnotations().stream()
+                .map(annotation -> new AnnotationWithTextSelection(annotation, createHighlighter(annotation)))
+                .toList());
+    }
+
+    public void registerAnnotationsUpdatedListener(Consumer<List<AnnotationWithTextSelection>> listener) {
+        annotationsUpdatedListener.add(listener);
+    }
+
+    public GradingConfig getGradingConfig() {
+        return assessment.getConfig();
+    }
+
+    public ClonedProgrammingSubmission getClonedSubmission() {
+        return clonedSubmission;
+    }
+
+    public void addAnnotationAtCaret(MistakeType mistakeType, boolean withCustomMessage) {
+        if (assessment == null) {
+            throw new IllegalStateException("No active assessment");
+        }
+
+        var selection = CodeSelection.fromCaret();
+        if (selection.isEmpty()) {
+            ArtemisUtils.displayGenericErrorBalloon("No code selected");
+            return;
+        }
+        var selectedText = selection.get().text();
+
+        var editor = EditorUtil.getActiveEditor();
+        int startLine = editor.getDocument().getLineNumber(selectedText.getStartOffset());
+        int endLine = editor.getDocument().getLineNumber(selectedText.getEndOffset());
+        String path = selection.get().projectRelativePath().toString();
+
+        Annotation annotation;
+        if (mistakeType.isCustomAnnotation()) {
+            double customScore = -1.0;
```

Comment:
Why hard coded score?

Reply:
Because I don't yet know what's the best way to open a number input window :D

Reply:
So probably add a TODO that we don't miss it 😂